ioctl/test_ioctl.c

```diff
@@ -0,0 +1,405 @@
+						TEST_MESSAGE("lseek failed in TIOCINOUT request");
+					}
+					if (write(fd, in_data, 4) != 4) {
+						TEST_MESSAGE("write failed in TIOCINOUT request");
+					}
+					out = 18;
+					out_data = (const void *)(&out);
+					break;
+				case TIOCINOUT_BIG:
+					if (lseek(fd, 0, SEEK_SET) != 0) {
+						TEST_MESSAGE("lseek failed in TIOCINOUT_BIG request");
+					}
+					if (write(fd, in_data, BIG_SIZE) != BIG_SIZE) {
+						TEST_MESSAGE("write failed in TIOCINOUT_BIG request");
+					}
+					memset(out2, 8, BIG_SIZE);
+					out_data = (const void *)(out2);
+					break;
+				default:
+					err = -1;
+					break;
+			}
+			ioctl_setResponse(&msg, request, err, out_data);
+		}
+		msgRespond(port, &msg, rid);
+	}
+	portDestroy(port);
+	return retval;
+}
+
+
+TEST_GROUP(ioctl);
+
+
+TEST_SETUP(ioctl)
+{
+}
+
+TEST_TEAR_DOWN(ioctl)
+{
+}
+
+
+TEST(ioctl, invalid_req)
+{
+	int ret = ioctl(dd, 0x1, NULL);
+	TEST_ASSERT_EQUAL_INT(-1, ret);
+}
+
+
+TEST(ioctl, regular_file)
+{
+	/* try regular file, not special device file */
+	int fd = open(PATH_REG, O_RDWR | O_CREAT | O_TRUNC, S_IFREG);
+	TEST_ASSERT_NOT_EQUAL_INT(-1, fd);
+	errno = 0;
+	int ret = ioctl(fd, TIOCSIG, NULL);
+	TEST_ASSERT_EQUAL_INT(-1, ret);
+	close(fd);
+	remove(PATH_REG);
+}
+
+
+TEST(ioctl, not_valid_fd)
+{
+	errno = 0;
+	int ret = ioctl(1234, 0, NULL);
+	TEST_ASSERT_EQUAL_INT(-1, ret);
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+}
+
+
+TEST(ioctl, no_data)
+{
+	int32_t flag;
+	int ret;
+
+	/* Send data to driver by value */
+	ret = ioctl(dd, TIOCSIG, NULL);
+	TEST_ASSERT_EQUAL_INT(0, ret);
+	TEST_ASSERT_EQUAL_INT(0, lseek(fd, 0, SEEK_SET));
+	TEST_ASSERT_EQUAL_INT(4, read(fd, &flag, 4));
+	TEST_ASSERT_EQUAL_INT32(1, flag);
+}
+
+
+TEST(ioctl, in_val)
+{
+	int32_t rdata = 0;
+	int32_t data = 14;
+	int ret;
+
+	/* Send data to driver by value */
+	ret = ioctl(dd, TIOCVAL, data);
+	TEST_ASSERT_EQUAL_INT(0, ret);
+	TEST_ASSERT_EQUAL_INT(0, lseek(fd, 0, SEEK_SET));
+	TEST_ASSERT_EQUAL_INT(4, read(fd, &rdata, 4));
+	TEST_ASSERT_EQUAL_INT32(14, rdata);
+}
+
+
+TEST(ioctl, data_in)
+{
+	int32_t rdata = 0;
+	int32_t data_in = 20;
+	int ret;
+
+	/* Send data to driver by pointer */
+	ret = ioctl(dd, TIOCPTR, &data_in);
+	TEST_ASSERT_EQUAL_INT(0, ret);
+	TEST_ASSERT_EQUAL_INT(0, lseek(fd, 0, SEEK_SET));
+	TEST_ASSERT_EQUAL_INT(4, read(fd, &rdata, 4));
+	TEST_ASSERT_EQUAL_INT32(20, rdata);
+}
+
+
+TEST(ioctl, data_in_big)
+{
+	/* Big data so that it is not copied inside message */
+	int8_t data_in[BIG_SIZE];
+	int8_t rdata[BIG_SIZE];
+	int ret;
+
+	TEST_ASSERT_EQUAL_PTR(data_in, memset(data_in, 3, BIG_SIZE));
+	TEST_ASSERT_EQUAL_PTR(rdata, memset(rdata, 0, BIG_SIZE));
+
+	/* Send data to driver by pointer , data big enough to not be copied by ioctl_pack directly into message */
+	ret = ioctl(dd, TIOCPTR_BIG, data_in);
+	TEST_ASSERT_EQUAL_INT(0, ret);
+	TEST_ASSERT_EQUAL_INT(0, lseek(fd, 0, SEEK_SET));
+	TEST_ASSERT_EQUAL_INT(BIG_SIZE, read(fd, rdata, BIG_SIZE));
```

Comment:
please note that read can return partial data according to specs (unlikely in this case, but to be 100% correct you should read in a loop).

No need to change anything for the sake of readability of the test.